txnsync/encodedgroupsunmarshalers.go

```diff
@@ -0,0 +1,795 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/transactions"
+)
+
+var errDataMissing = errors.New("failed to decode: data missing")
+
+func getSlice(b []byte, index int, size int) ([]byte, error) {
+	if index*size+size > len(b) {
+		return nil, errDataMissing
+	}
+	return b[index*size : index*size+size], nil
+}
+
+func getNibble(b []byte, index int) (byte, error) {
+	if index > len(b)*2 {
+		return 0, errDataMissing
+	}
+	if index%2 == 0 {
+		return b[index/2] / 16, nil
+	}
+	return b[index/2] % 16, nil
+}
+
+func (stub *txGroupsEncodingStub) reconstructSignedTransactions(signedTxns []transactions.SignedTxn, genesisID string, genesisHash crypto.Digest) error {
+	var index int
+	index = 0
+	for i := range signedTxns {
+		if exists := stub.BitmaskSig.EntryExists(i, int(stub.TotalTransactionsCount)); exists {
+			slice, err := getSlice(stub.Sig, index, len(crypto.Signature{}))
+			if err != nil {
+				return err
+			}
+			copy(signedTxns[i].Sig[:], slice)
+			index++
+		}
+	}
+	if err := stub.reconstructMsigs(signedTxns); err != nil {
+		return fmt.Errorf("failed to msigs: %w", err)
+	}
+	if err := stub.reconstructLsigs(signedTxns); err != nil {
+		return fmt.Errorf("failed to lsigs: %w", err)
+	}
+	index = 0
+	for i := range signedTxns {
+		if exists := stub.BitmaskAuthAddr.EntryExists(i, int(stub.TotalTransactionsCount)); exists {
+			slice, err := getSlice(stub.AuthAddr, index, crypto.DigestSize)
+			if err != nil {
+				return err
+			}
+			copy(signedTxns[i].AuthAddr[:], slice)
+			index++
+		}
+	}
```

Comment:
I was reviewing the executing of this section - in some cases it would be extremely inefficient:
let's say you have a 500 signed transactions. And there is only a single transaction with AuthAddr.
In that case, you would perform this loop 500 times, calling EntryExists over and over.

The better implementation for this would be to use the bitmask to figure out what the indices are, and then go back and fill up the data accordingly. This shouldn't be too difficult to implement -
have the bitmap expose a Iterate method that makes a callback to an anonymous function.
i.e.
```golang
err := stub.BitmaskAuthAddr.Iterate(stub.TotalTransactionsCount, func (i int) error {
     slice, err := getSlice(stub.AuthAddr, index, crypto.DigestSize)
     if err != nil {
	return err
     }
     copy(signedTxns[i].AuthAddr[:], slice)
     index++
})
if err != nil {
    return err
}
```